xxxx-ledger-plugin-interface-2/xxxx-ledger-plugin-interface-2.md

```diff
@@ -0,0 +1,502 @@
+
+Create a new instance of the plugin. Each instance typically corresponds to a different ledger. However, some plugins MAY deviate from a strict one-to-one relationship and MAY use one instance for multiple (similar) ledgers or multiple instances to talk to the same ledger.
+
+Throws `InvalidFieldsError` if the constructor is given incorrect arguments.
+
+###### Parameters
+| Name | Type | Description |
+|:--|:--|:--|
+| opts | <code>[PluginOptions](#class-pluginoptions)</code> | Object containing ledger-related settings. May contain plugin-specific fields. |
+
+###### Example
+```js
+const ledgerPlugin = new LedgerPlugin({
+
+  // auth parameters are defined by the plugin
+
+  _store: {
+    // persistence may be required for internal use by some ledger plugins
+    // (e.g. when the ledger has reduced memo capability and we can only put an ID in the memo)
+    // Store a value under a key
+    put: (key, value) => {
+      // Returns Promise.<null>
+    },
+    // Fetch a value by key
+    get: (key) => {
+      // Returns Promise.<Object>
+    },
+    // Delete a value by key
+    del: (key) => {
+      // Returns Promise.<null>
+    }
+  }
+})
+```
+
+For a detailed description of these properties, please see [`PluginOptions`](#class-pluginoptions).
+
+### Connection Management
+
+#### LedgerPlugin#connect
+<code>ledgerPlugin.connect( options:[ConnectOptions](#class-connectoptions ) ⇒ Promise.&lt;null></code>
+
+`options` is optional.
+
+Initiate ledger event subscriptions. Once `connect` is called the ledger plugin MUST attempt to subscribe to and report ledger events. Once the connection is established, the ledger plugin should emit the [`connect`](#event-connect-) event. If the connection is lost, the ledger plugin SHOULD emit the [`disconnect`](#event-disconnect-) event. The plugin should ensure that the information returned by `getInfo` is available and cached before emitting the [`connect`](#event-connect-) event.
+
+Rejects with `InvalidFieldsError` if credentials are missing, and `NotAcceptedError` if credentials are rejected.
+Rejects with `TypeError` if `options.timeout` is passed but is not a `Number`.
+
+#### LedgerPlugin#disconnect
+<code>ledgerPlugin.disconnect() ⇒ Promise.&lt;null></code>
+
+Unsubscribe from ledger events.
+
+#### LedgerPlugin#isConnected
+<code>ledgerPlugin.isConnected() ⇒ Boolean</code>
+
+Query whether the plugin is currently connected.
+
+#### LedgerPlugin#getInfo
+<code>ledgerPlugin.getInfo() ⇒ [LedgerInfo](#class-ledgerinfo)</code>
+
+Retrieve some metadata about the ledger. Throws `NotConnectedError` if the plugin is not connected to the ledger.
+
+###### Example Return Value
+```json
+{
+  "currencyCode": "USD",
+  "currencyScale": 4
+}
+```
+
+For a detailed description of these properties, please see [`LedgerInfo`](#class-ledgerinfo).
+
+#### Event: `connect`
+<code>ledgerPlugin.on('connect', () ⇒ )</code>
+
+Emitted whenever a connection is successfully established.
+
+#### Event: `disconnect`
```

Comment:
> Should ledger plugins try to reconnect themselves or should that be up to the user of the plugin?

-1 - That seems like functionality that should be written once in whatever consumes the plugin.

Also, a plugin CAN try to reconnect silently if it is able to buffer requests in the interim. That would be a feature of a good plugin but seems like a high bar for all plugins.

Finally, that makes some assumptions about the underlying transport that will be used which I think is a mistake.

Reply:
Currently, the connector listens to the `disconnect` event to know when it should stop advertising a certain route. In fact, it actively informs peers that it can no longer service this route.

For this use case alone, I think this is an important feature.